### PR TITLE
[MMP] Allow resolving assemblies to the ones passed in command line args

### PR DESCRIPTION
**Pull request by Therzok**
This reverts the recursive directory search commit and allows resolving assemblies from the ones passed.

### CODE REVIEW

**Comment by Therzok** [comment]
I think this should be cherry-picked to 15.7, so the --recursive-directories option is not surfaced to anyone.

**Comment by chamons** [comment]
I agree on removing --recursive-directories in 15.7. I am hesitant to add the 
```
+                CommandLineAssemblies = RootAssemblies,
```
bit since that changes behavior, and mmp behavior is non-trivial and not testing as much as I'd like. 

Though it is likely fine.

**Comment by chamons** [comment]
build

File: tools/common/Application.cs [hunk 1]
```diff
@@ -524,10 +524,10 @@ public void RunRegistrar ()
 			var resolver = new PlatformResolver () {
 				FrameworkDirectory = Driver.GetPlatformFrameworkDirectory (this),
 				RootDirectory = Path.GetDirectoryName (RootAssembly),
-			};
 #if MMP
-			resolver.RecursiveSearchDirectories.AddRange (Driver.RecursiveSearchDirectories);
+				CommandLineAssemblies = RootAssemblies,
```
Comment (Therzok) [inline tools/common/Application.cs]: This is what actually makes the MonoMacResolver actually resolve to assemblies given as arguments. I can add another list to be checked, but I think this is fine.

1. This code path is `--runregistrar` specific
2. The other bit of code which sets this is inside `Pack`, which is the _other_ hit when the action is not `--runregistrar`

https://github.com/xamarin/xamarin-macios/blob/3113c5d2b5cca730372177e9b1cee17062a959ff/tools/mmp/driver.cs#L513

Comment (chamons) [reply tools/common/Application.cs]: I agree that this is the correct fix at face value. And I'm totally cool with it going into master.

I've just been burned enough times with last minute behavior changes in release branches that i'm hesitant.

Comment (Therzok) [reply tools/common/Application.cs]: Want me to split the PR into 2 commits? One which reverts recursive search dirs and one which implements this?

So we can backport the revert safely?

Comment (chamons) [reply tools/common/Application.cs]: That would be grand. ❤️

Comment (Therzok) [reply tools/common/Application.cs]: Done.

**Comment by spouliot** [comment]
Note: this *must* be in `d15-7` otherwise the option will become public (and can't be removed as easily).

**Comment by spouliot** [comment]
@chamons I think that before (unreleased) 15.7 it was not allowed to have more than one root assembly - so whatever behaviour is new (and can't break existing apps), right ?

**Comment by chamons** [comment]
I _think_ so @spouliot but I'm not 100% certain without some archaeology.

**Comment by rolfbjarne** [comment]
This change is only when passing --runregistrar to mmp, which is an internal option: https://github.com/xamarin/xamarin-macios/blob/040461ba0ac9a6cf8ffa605b5f9998e612579324/tools/mmp/driver.cs#L355-L360

Thus we shouldn't have to care for backwards compatibility.

**Comment by chamons** [comment]
Alright, then I can stop being paranoid (in this case).

**Comment by chamons** [comment]
Strange...

```
17:28:54.0766340 ERROR [2018-02-22 17:28:54Z]: Error reading assembly '/Library/Frameworks/Mono.framework/Versions/5.10.0/lib/mono/4.0-api/System.DirectoryServices.dll' in framework '.NETFramework,Version=v4.0':
17:28:54.0766830 System.Threading.ThreadAbortException
```
```
Assertion failure in void assertRunningOnAppKitThread
```

Trying one more time before filing maccore issues.

**Comment by chamons** [comment]
build

**Comment by chamons** [comment]
build

**Comment by chamons** [comment]
Test failures are random and knwon:
https://github.com/xamarin/maccore/issues/658
https://github.com/xamarin/maccore/issues/592

**Comment by chamons** [comment]
https://github.com/xamarin/xamarin-macios/pull/3651 for d15-7

**Pull request merged**